### PR TITLE
Fix past_present_share_buffer=false kv cache bug and asserts

### PR DESCRIPTION
**Pull request by aciddelgado**
This PR fixes a bug where, when using continuous decoding, the output would differ depending on whether past_present_share_buffer was on or off. The bug was actually deeper than this, as the KV-Cache was missing the last generated token when using continuous decoding, regardless of share buffer.

It also fixes a few assert statements.

### CODE REVIEW

File: src/generators.cpp [hunk 1]
```diff
@@ -283,6 +283,12 @@ void Generator::AppendTokens(const cpu_span<int32_t>& input_ids) {
   if (search_->GetSequenceLength() != 0 && state_->params_->search.batch_size > 1)
     throw std::runtime_error("AppendTokens can only be called once for batch_size > 1. To call AppendTokens again, use RewindToLength(0)");
 
+  if (just_generated_) {
+    just_generated_ = false;
+    auto next_tokens = search_->GetNextTokens();
+    ComputeLogits(next_tokens);
```
Comment (RyanUnderhill) [inline src/generators.cpp:289]:
```suggestion
    ComputeLogits(search_->GetNextTokens());
```

File: src/generators.h [hunk 2]
```diff
@@ -129,7 +129,8 @@ struct Generator : LeakChecked<Generator> {
  private:
   DeviceSpan<int32_t> AllocateInputIdsOnDevice(const cpu_span<int32_t>& input_ids);
   void ComputeLogits(DeviceSpan<int32_t>& next_tokens);
-  bool just_rewinded_{false};
+  bool just_rewinded_{};
```
[the remark this answers is not in the document]
Comment (RyanUnderhill) [reply src/generators.h:132]: Then would it make more sense to have an enum of the states?
enum State { normal, rewound, generated };

(you can probably think of a better name for normal)

**Comment by guschmue** [comment]
tested with webgpu and dynamic kv_cache ... super, that works now

**Review by guschmue — APPROVED** [review]

File: src/generators.h [hunk 3]
```diff
@@ -127,9 +127,12 @@ struct Generator : LeakChecked<Generator> {
   bool computed_logits_{};  // Set to true in ComputeLogits() and false after appending a token to ensure a 1 to 1 call ratio
 
  private:
-  DeviceSpan<int32_t> AllocateInputIdsOnDevice(const cpu_span<int32_t>& input_ids);
-  void ComputeLogits(DeviceSpan<int32_t>& next_tokens);
-  bool just_rewinded_{false};
+  DeviceSpan<int32_t> AllocateInputIdsOnDevice(const cpu_span<int32_t> input_ids);
+  void ComputeLogits(DeviceSpan<int32_t> next_tokens);
+  enum Action { standard,
```
Comment (RyanUnderhill) [inline src/generators.h:132]: Looks good, can you add some one line comments explaining each state?

**Pull request merged**